ctk/idp/src/main/kotlin/org/codice/compliance/web/slo/endpoint/ServiceProviderEndpointVerification.kt

```diff
@@ -0,0 +1,151 @@
+import org.apache.cxf.rs.security.saml.sso.SSOConstants.SAML_RESPONSE
+import org.apache.cxf.rs.security.saml.sso.SSOConstants.SIGNATURE
+import org.apache.cxf.rs.security.saml.sso.SSOConstants.SIG_ALG
+import org.apache.wss4j.common.saml.OpenSAMLUtil
+import org.codice.compliance.Common.Companion.getSingleLogoutLocation
+import org.codice.compliance.utils.TestCommon.Companion.RESPONDER
+import org.codice.compliance.utils.TestCommon.Companion.SAML_ENCODING
+import org.codice.compliance.utils.TestCommon.Companion.SUCCESS
+import org.codice.compliance.utils.TestCommon.Companion.currentSPIssuer
+import org.codice.compliance.utils.TestCommon.Companion.samlObjectToString
+import org.codice.compliance.utils.sign.SimpleSign
+import org.codice.compliance.verification.binding.PostBindingVerifier
+import org.codice.compliance.verification.binding.PostFormVerifier
+import org.codice.compliance.verification.binding.RedirectBindingVerifier
+import org.codice.compliance.verification.core.requests.CoreLogoutRequestProtocolVerifier
+import org.codice.security.saml.SamlProtocol
+import org.codice.security.sign.Encoder
+import org.joda.time.DateTime
+import org.opensaml.saml.common.SAMLVersion
+import org.opensaml.saml.saml2.core.LogoutRequest
+import org.opensaml.saml.saml2.core.impl.IssuerBuilder
+import org.opensaml.saml.saml2.core.impl.LogoutResponseBuilder
+import org.opensaml.saml.saml2.core.impl.StatusBuilder
+import org.opensaml.saml.saml2.core.impl.StatusCodeBuilder
+import java.net.URLDecoder
+import java.nio.charset.StandardCharsets
+import java.util.UUID
+import javax.ws.rs.core.UriBuilder
+import kotlin.collections.set
+
+class ServiceProviderEndpointVerification(private val sendValidResponse: Boolean = true) {
+
+    private var relayState: String? = null
+    private val logoutResponse = LogoutResponseBuilder().buildObject().apply {
+        id = "a" + UUID.randomUUID().toString()
+        version = SAMLVersion.VERSION_20
+        issueInstant = DateTime()
+        issuer = IssuerBuilder().buildObject().apply { value = currentSPIssuer }
+        status = StatusBuilder().buildObject().apply {
+            StatusCodeBuilder().buildObject().apply {
+                value = if (sendValidResponse) SUCCESS else RESPONDER
+            }
+        }
+    }
+
+    /* Validations */
+
+    internal fun redirectRequestIsValid(queryParameters: Parameters): Boolean {
+        val paramMap = mutableMapOf<String, String>()
+        val request = queryParameters[SAML_REQUEST] ?: return false
+
+        paramMap[SAML_REQUEST] = request
+        queryParameters[SAML_ENCODING]?.let { paramMap[SAML_ENCODING] = it }
+        queryParameters[SIGNATURE]?.let { paramMap[SIGNATURE] = it }
+        queryParameters[SIG_ALG]?.let { paramMap[SIG_ALG] = it }
+        queryParameters[RELAY_STATE]?.let {
+            RedirectBindingVerifier().verifyRelayStateLength(it)
+            paramMap[RELAY_STATE] = it
+            relayState = it
+        }
```

Comment:
❗️ We need to verify that all of these (minus relay state) are **not** null. See `RedirectBindingVerifier.verifyNoNulls`. 👈 that method could be reused if we change the method to just take a url string and if we change this SPVerifier method to take the URL instead of query parameters.

Reply:
Looking at this further, I'm not sure what verifications I need to add, `verifyNoNullsAndParse` verifies:
1- the location header, which won't exist since the redirect happens before the SP2 endpoint is hit
2- the location url, which is verified by the fact that the second SP endpoint `/services/samldsa/logout/redirect` is hit
3- the fact that there are parameters, which is verified by checking that `queryParameters[SAML_REQUEST]` is not null
4- that the response (in this case request) is not null, which is verified by ☝️ 
5- checks that relay state is returned, which doesn't apply.

Reply:
After doing more research into this method, it is good as is. All of the nulls that could happen are either allowed to be null, or are properly checked in the called methods.